File: Power Query/400-499/405/PQ_405.py
import pandas as pd

path = "400-499/405/PQ_Challenge_405.xlsx"
input = pd.read_excel(path, usecols="A:D", nrows=25, skiprows=0)
test = pd.read_excel(path, usecols="F:O", nrows=6, skiprows=0)

m = (
    input.loc[input.index.repeat(2), ["MatchID"]]
    .reset_index(drop=True)
    .assign(
        Team=input[["HomeTeam", "AwayTeam"]].stack().to_numpy(),
        GF=input["Score"].str.split("-").explode().astype(int).to_numpy(),
    )
)
m["GA"] = m.groupby("MatchID")["GF"].transform("sum") - m["GF"]
result = (
    m.assign(W=m.GF.gt(m.GA), D=m.GF.eq(m.GA), L=m.GF.lt(m.GA))
    .groupby("Team", as_index=False)
    .agg(
        Played=("GF", "size"),
        Wins=("W", "sum"),
        Draws=("D", "sum"),
        Losses=("L", "sum"),
        GF=("GF", "sum"),
        GA=("GA", "sum"),
    )
    .assign(GD=lambda d: d.GF - d.GA, Points=lambda d: d.Wins * 3 + d.Draws)
    .sort_values(
        ["Points", "GD", "GF", "Team"],
        ascending=[False] * 3 + [True],
        ignore_index=True,
    )
)
result.insert(0, "Rank", result.index + 1)

print(result.equals(test))
# True

File: Power Query/400-499/405/PQ_405.R
library(tidyverse)
library(readxl)

path <- "400-499/405/PQ_Challenge_405.xlsx"
input <- read_excel(path, range = "A1:D25")
test <- read_excel(path, range = "F1:O7")

result <- input %>%
  unite("pair", HomeTeam, AwayTeam, sep = "-") %>%
  separate_longer_delim(c(pair, Score), delim = "-") %>%
  mutate(Score = as.numeric(Score)) %>%
  mutate(
    Team = pair,
    GF = Score,
    GA = rev(Score),
    .by = MatchID
  ) %>%
  summarise(
    Played = n(),
    Wins = sum(GF > GA),
    Draws = sum(GF == GA),
    Losses = sum(GF < GA),
    GF = sum(GF),
    GA = sum(GA),
    .by = Team
  ) %>%
  mutate(
    GD = GF - GA,
    Points = Wins * 3 + Draws,
    Rank = min_rank(desc(Points))
  ) %>%
  arrange(desc(Points), desc(GD), desc(GF), Team) %>%
  mutate(Rank = row_number()) %>%
  select(Rank, Team, Played, Wins, Draws, Losses, GF, GA, GD, Points)

all.equal(result, test)
# True
